safe_control_gym/controllers/mpc/mpc_acados.py

```diff
@@ -0,0 +1,391 @@
+        ocp.cost.Vu = np.zeros((ny, nu))
+        ocp.cost.Vu[nx:(nx + nu), :nu] = np.eye(nu)
+        ocp.cost.Vx_e = np.eye(nx)
+        # placeholder y_ref and y_ref_e (will be set in select_action)
+        ocp.cost.yref = np.zeros((ny, ))
+        ocp.cost.yref_e = np.zeros((ny_e, ))
+
+        # Constraints
+        # general constraint expressions
+        state_constraint_expr_list = []
+        input_constraint_expr_list = []
+        for sc_i, state_constraint in enumerate(self.state_constraints_sym):
+            state_constraint_expr_list.append(state_constraint(ocp.model.x))
+        for ic_i, input_constraint in enumerate(self.input_constraints_sym):
+            input_constraint_expr_list.append(input_constraint(ocp.model.u))
+
+        h_expr_list = state_constraint_expr_list + input_constraint_expr_list
+        h_expr = cs.vertcat(*h_expr_list)
+        h0_expr = cs.vertcat(*h_expr_list)
+        he_expr = cs.vertcat(*state_constraint_expr_list)  # terminal constraints are only state constraints
+        # pass the constraints to the ocp object
+        ocp = self.processing_acados_constraints_expression(ocp, h0_expr, h_expr, he_expr)
+
+        # slack costs for nonlinear constraints
+        if self.soft_constraints:
+            # slack variables for all constraints
+            ocp.constraints.Jsh_0 = np.eye(h0_expr.shape[0])
+            ocp.constraints.Jsh = np.eye(h_expr.shape[0])
+            ocp.constraints.Jsh_e = np.eye(he_expr.shape[0])
+            # slack penalty
+            L2_pen = self.soft_penalty
+            L1_pen = self.soft_penalty
+            ocp.cost.Zl_0 = L2_pen * np.ones(h0_expr.shape[0])
+            ocp.cost.Zu_0 = L2_pen * np.ones(h0_expr.shape[0])
+            ocp.cost.zl_0 = L1_pen * np.ones(h0_expr.shape[0])
+            ocp.cost.zu_0 = L1_pen * np.ones(h0_expr.shape[0])
+            ocp.cost.Zu = L2_pen * np.ones(h_expr.shape[0])
+            ocp.cost.Zl = L2_pen * np.ones(h_expr.shape[0])
+            ocp.cost.zl = L1_pen * np.ones(h_expr.shape[0])
+            ocp.cost.zu = L1_pen * np.ones(h_expr.shape[0])
+            ocp.cost.Zl_e = L2_pen * np.ones(he_expr.shape[0])
+            ocp.cost.Zu_e = L2_pen * np.ones(he_expr.shape[0])
+            ocp.cost.zl_e = L1_pen * np.ones(he_expr.shape[0])
+            ocp.cost.zu_e = L1_pen * np.ones(he_expr.shape[0])
+
+        # placeholder initial state constraint
+        x_init = np.zeros((nx))
+        ocp.constraints.x0 = x_init
+
+        # set up solver options
+        ocp.solver_options.qp_solver = 'PARTIAL_CONDENSING_HPIPM'
+        ocp.solver_options.hessian_approx = 'GAUSS_NEWTON'
+        ocp.solver_options.integrator_type = 'DISCRETE'
+        ocp.solver_options.nlp_solver_type = 'SQP' if not self.use_RTI else 'SQP_RTI'
+        ocp.solver_options.nlp_solver_max_iter = 25 if not self.use_RTI else 1
+        # ocp.solver_options.globalization = 'FUNNEL_L1PEN_LINESEARCH' if not self.use_RTI else 'MERIT_BACKTRACKING'
```

Comment:
as usual, i think commented out code needs to be either deleted or included unless it really serves a purpose

Reply:
I have this for two reasons 1) for the user to know there are some options to improve the MPC solving; 2) easy for me to try out when needed. So I would like to keep them if this doesn't matter that much.